portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -39,26 +44,40 @@
   private List<Comment> arr = new ArrayList<>();
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
     Query commentsQuery = new Query("Comment").addSort("timestamp");
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
     PreparedQuery results = datastore.prepare(commentsQuery);
     List<Comment> arr = new ArrayList<>();
     int how_many = Integer.parseInt(request.getParameter("how_many"));
+    String langCode = (String) request.getParameter("lang");
 
+    String[] acceptedCodes = {"en", "ro", "de", "fr"};
+    if (!Arrays.asList(acceptedCodes).contains(langCode)){
+      response.setContentType("text/html");
+      response.getWriter().print("<script>alert(\"Please use an accepted language code.\")</script>");
```

Comment:
I see two issues with this approach. 

1. Consistency of server and client code: The list of valid language codes is maintained in the html file and in the server code. I don't think there is a straightforward solution the way the project is set up right now.
2. The message that is shown to the user: For debugging, it would be good to see a list of supported languages.

I would also suggest to change the message to be more user-friendly, to something like "This language is currently not supported. Supported languages are: ..."

Reply:
Besides the comment on the user-friendly message, I don't really get what you are saying. There is no method that returns a list of valid languages and I just selected four languages, which are in both lists, from servlet and from html. I have the list in both of them because: I needed the list in html to be able to send the post request and I also need it in servlet to be able to verify that the post request in valid and a user doesn't try to break the code by sending another parameter for the language code.

Reply:
> There is no method that returns a list of valid languages
If that is the case then the current implementation is ok.

> I needed the list in html to be able to send the post request and I also need it in servlet to be able to verify that the post request in valid and a user doesn't try to break the code by sending another parameter for the language code.
You could define the list of accepted languages once in the server and retrieve it to build the client-side list. But as I said, the current implementation is ok as-is.